src/jit/morph.cpp

```diff
@@ -2887,24 +2887,38 @@ GenTreeCall* Compiler::fgMorphArgs(GenTreeCall* callNode)
         }
         else
 #endif // defined(_TARGET_X86_)
-        if (call->gtCallType == CT_INDIRECT && call->gtCallCookie)
+        if (call->gtCallType == CT_INDIRECT && (call->gtCallCookie != nullptr))
         {
             assert(!call->IsUnmanaged());
 
             GenTree* arg = call->gtCallCookie;
             noway_assert(arg != nullptr);
             call->gtCallCookie = nullptr;
 
-            call->gtCallArgs = gtNewListNode(arg, call->gtCallArgs);
-            numArgs++;
+#if defined(_TARGET_X86_)
+            // x86 passes the cookie on the stack as the final argument to the call.
+            GenTreeArgList* node = gtNewListNode(arg, nullptr);
+            if (call->gtCallArgs == nullptr)
+            {
+                call->gtCallArgs = node;
+            }
+            else
+            {
+                GenTreeArgList* insertionPoint = call->gtCallArgs;
+                for (; insertionPoint->Rest() != nullptr; insertionPoint = insertionPoint->Rest())
+                {
+                }
 
-            // x86 passes the cookie on the stack.
-            CLANG_FORMAT_COMMENT_ANCHOR;
+                insertionPoint->Rest() = node;
+            }
+
```

Comment:
I'm surprised this works. The nonStandardArgs mechanism was only designed for register arguments. If you look at where it is used, you can see it is only used when morphing register arguments. Maybe all you needed to do is add an argument and mark it REG_STK (instead of the default REG_NA?), and presumably that is sufficient for fgMorphArgs() to not try to put it in a register (in this case, presumably it would try to stick it in EDX?) I don't think you should add it to the nonStandardArgs table, since that doesn't do anything.

---
In reply to: [84548962](https://github.com/dotnet/coreclr/pull/7758#discussion_r84548962) [](ancestors = 84548962)